Feature/dotnet 6 migration

BenvdStouwe opened the pull request:
Hi Matty, I migrated your builderbuilder project to dotnet 6 (still scoped to Windows due to Forms), migrated the tests to xunit en let my editor perform a lot of automatic refactors. This made the PR a bit large, sorry for that. 

The tests are passing and the application still works (on my machine 😉).

I also made a new compiler that takes a class as input and generates a record that you can manipulate with the `with` operator instead of all the `With` methods. That is not in this PR but I can create a new one for it if you like. 

I also plan to build a source code generator that generates testbuilders on the go, that'll have to be a huge refactor and I don't know when (and if...) that will land.

BenvdStouwe committed:
Migrate to dotnet 6 and convert tests to xunit

BenvdStouwe committed:
Automatic code fixes

BenvdStouwe committed:
Automatic fixes

Mattias1 commented on BuilderBuilder.Test/AssertHelper.cs:
```diff
@@ -1,86 +1,73 @@
-        public static void ListEq<T, C>(IEnumerable<T> result, Func<T, C> compareBy, IEnumerable<C> expected) {
-            ListEq(result.Select(compareBy), expected);
+    private static void ListEq<T>(IEnumerable<T> result, IEnumerable<T> expected) {
+        var actualArr = result as T[] ?? result.ToArray();
+        var expectedArr = expected as T[] ?? expected.ToArray();
```
What's the difference here with `result.ToArray();`? Or am I being naive in thinking that `IEnumerable.ToArray()` on an actual array will just return `this` and does it loop over the entire array in order to create itself again?

And even if it is faster (I'm assuming this is a performance optimization), what's the benefit of working with a raw array? We use lots of link expressions later on, which do not benefit at all from arrays.

BenvdStouwe replied:
The `ToArray` makes sure the enumerable is only enumerated once, since the parameter is used multiple times in this method. Usually multiple enumerations shouldn't change the result of the enumeration, but I've seen cases where it did so I make sure they are always enumerated once. 

The `as T[]` looks like a micro-optimization that I will remove.

Mattias1 commented on BuilderBuilder.Test/AssertHelper.cs:
```diff
@@ -1,86 +1,73 @@
-            return $"{description}: {errors}";
+    private static string ErrorList<T>(string description, IEnumerable<T> list) {
+        var arr = list as T[] ?? list.ToArray();
+        if (!arr.Any()) {
```
Is this an optimization attempt as well? Because this feels counter productive to me if it is.

`.Any()` is a constant operation, it either contains something, or it doesn't.
The `.ToArray()` here loops over the entire list, which is not a constant operation and therefore slower.

Furthermore, later in the code we use `.Select` and we're not making use of the internal array anymore.

BenvdStouwe replied:
On an array it is a constant operation, but this is an IEnumerable so it has to start enumerating before it can know that there is any value. I changed the parameters of the private methods to IList so there is no chance for multiple enumerations.

Mattias1 commented on BuilderBuilder.Test/Compilers/ImmutableCompilerTest.cs:
```diff
@@ -1,30 +1,32 @@
+        var input = new BuilderEntity(persistable: false)
+        {
+            Name = "ExampleEntity"
+        };
```
This is probably an automatic change, but I fail to see why this is better.
If you added all properties this way sure, but the fields are still set manually below.

I'd say either roll this back, or also include the fields in this style.

Mattias1 commented on BuilderBuilder.Test/Compilers/VipCompilerTest.cs:
```diff
@@ -1,30 +1,32 @@
+        var input = new BuilderEntity(persistable: true)
+        {
+            Name = "ExampleEntity"
+        };
```
Same as above.

Mattias1 commented on BuilderBuilder.Test/Compilers/VipCompilerTest.cs:
```diff
@@ -97,25 +99,26 @@ public override ExampleEntity AutoBuild()
+        var input = new BuilderEntity(persistable: false)
+        {
+            Name = "ExampleEntity"
+        };
```
^^

Mattias1 commented on BuilderBuilder/BuilderEntity.cs:
```diff
@@ -1,34 +1,31 @@
+    public string Name { get; set; }
 
-        public List<Field> Fields { get; private set; }
+    public List<Field> Fields { get; }
```
Thanks

Mattias1 commented on BuilderBuilder/Compilers/Compiler.cs:
```diff
@@ -1,76 +1,74 @@
-    public abstract class Compiler
-    {
-        public virtual int TabSize => 4;
+    private static int TabSize => 4;
```
I'd argue that given this is an abstract class you may want to override this in a compiler if you want 2 spaces. But then, YAGNI. So I guess you're right here.

Mattias1 commented on BuilderBuilder/Compilers/ImmutableCompiler.cs:
```diff
@@ -1,102 +1,108 @@
+    private void AddExampleBuilder()
+    {
+        AddLine($"public static {BuilderClass} Builder()");
+        WithBlock(() => { AddLine($"return new {BuilderClass}();"); });
```
The nice thing about formatting this the old way is that it looks like the code this is generating - that also puts the `{` and `}` on new lines.

Mattias1 commented on BuilderBuilder/Compilers/ImmutableCompiler.cs:
```diff
@@ -1,102 +1,108 @@
+    private void AddFieldVariable(Field field)
+    {
+        var type = Nullable(field.Type);
+        var name = PrivateVar(field.Name);
```
So, code-style-purist technically `name` is far better than `_name` here. However, this code generates private variables with an underscore, so the `_name` is actually how it looks in the generated code. That's why I formatted it this way.

Mattias1 replied:
Some more thinking, I think you version is better. Leave it that way.

Mattias1 commented on BuilderBuilder/Compilers/ImmutableCompiler.cs:
```diff
@@ -1,102 +1,108 @@
+        var type = field.Type;
+        var name = field.Name;
+        var localVarName = LocalVar(field.Name);
+        var privatePropertyName = PrivateVar(field.Name);
```
Same as above: code-style-purist way this new code is better. However, the old code is much clearer in what the distinction is between `Name`, `name` and `_name`. Because that's exactly how they're displayed.

Mattias1 replied:
Thinking some more about this - if we want a compiler to not use underscores for example, than moving away from the literal mapping is better I guess. So maybe let's format it the official way instead of trying to be smart about it.

In that case though, `name` should be renamed to `publicPropertyName`.

Mattias1 commented on BuilderBuilder/Compilers/VipCompiler.cs:
```diff
@@ -1,106 +1,115 @@
-            AddLine($"public {BuilderClass}() : base() {{ }}");
-            AddEmptyLine();
+            if (field.InverseHandling == Field.InverseHandlingType.ManyToOne ||
+                field.InverseHandling == Field.InverseHandlingType.ManyToMany)
```
This is a personal preference, so feel free to ignore it, but could you put the `||` on the next line? That way you can see at the start of the lines the 'structure' of your check. In this case it doesn't matter much, but if you're ever running into an if like this, than it really helps to quickly see the structure of how your if works.

```
if ((aaaa
        && bbbb)
    || (aaaa
        && cccc)
    || (aaaa
        && bbbb
        && !cccc))
{
    // ....
}
```
Formatting ifs this way, my eyes only need to travel up and down to quickly understand how this if works. Moving sideways is only necessary to pick out the details.

Apart from the practical things, it's also more consistent with the C# programming language (and most others) in general - most operators come before the details. Probably for the same reason: structure left, meaning right.

Mattias1 reviewed (changes requested):
Hey Ben, good to see something from you ;)

Thanks for getting this up to dotnet 6 and xunit, much better this way.
I have plans to migrate my forms library to AvaloniaUI or MAUI, so that it runs on linux without needing something like wine, but not sure if I'll get to that this year ;)

If you have more compilers those are always welcome, so please make a PR (but a separate one).

I'm not very familiar with source code generators, but would this tool still make sense then? The premise being that you generate things once, and then have the freedom to tune your builders afterwards to your taste. Or would it better be suited for a different tool?

Anyway, for as far as the PR goes it looks good. I have a few small comments. Not sure howmuch time you have for notpicking, this PR is a huge improvement already so feel free to ignore most of them if you don't have the time.
The only one I feel strongly about is the comment in ImmutableCompiler.cs about renaming that variable to "publicPropertyName".

BenvdStouwe commented:
> Hey Ben, good to see something from you ;)
> 
> Thanks for getting this up to dotnet 6 and xunit, much better this way. I have plans to migrate my forms library to AvaloniaUI or MAUI, so that it runs on linux without needing something like wine, but not sure if I'll get to that this year ;)
> 
> If you have more compilers those are always welcome, so please make a PR (but a separate one).
> 
> I'm not very familiar with source code generators, but would this tool still make sense then? The premise being that you generate things once, and then have the freedom to tune your builders afterwards to your taste. Or would it better be suited for a different tool?
> 
> Anyway, for as far as the PR goes it looks good. I have a few small comments. Not sure howmuch time you have for notpicking, this PR is a huge improvement already so feel free to ignore most of them if you don't have the time. The only one I feel strongly about is the comment in ImmutableCompiler.cs about renaming that variable to "publicPropertyName".

Thanks for the feedback. I agree with a lot, sometimes I follow my editor a bit too much. I'll fix it later this week.

I think you're either going to love or hate source code generators: https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/source-generators-overview

It's code that will be generated compiletime (or "developtime" when your editor can handle it), you don't have the chance to manually change them. Using this you could generate builders for classes based on things like namespace, interfaces or attributes. When I would have a class like this:
```
public class Entity 
{
    public long Id { get; set; }
    public string Name { get; set; }
}
```
the generator could generate a record like this when compiling:
``` 
public record EntityBuilder 
{
    public long Id { private get; init; }
    public string Name { private get; init; }
   
    public Entity Build() => new() {
        Id = Id,
        Name = Name,
    }
}
```
and when you're writing code, your editor already knows that there will be an `EntityBuilder` and you can use it to setup a default implementation for tests:
```
public static class Entities
{
    public static EntityBuilder Default => new() 
    {
        Id = IdGenerator.Next
    }
}
````
Which you could use like this: `var entity = (Entities.Default with { Name = "Something meaningful" }).Build();`
and... it will(/should) compile. 

I haven't experimented with source generators yet but thought that this could be a nice intro for me.

Mattias1 commented:
Ah, that's definitely interesting and useful. Personally if I'd use that idea for test builders I'd use it to generate a builder base record, and then I can still tweak them by adding methods in a record that inherits from that generated record.
Best of both worlds :P

That said, I would not use that for this tool, but instead create a new library or something and then edit the readme here to point to that as an alternative.

Because using source generators would fix one major annoyance with these generated builders: if I change an entity later on, I have to edit my builder manually to include the new field.

BenvdStouwe commented:
Well you could always do that by making the record `partial` and making sure they are in the same namespace. Usually I really dislike `partial` but the projects that provide source generations use `partial` quite often.

BenvdStouwe committed:
Improve formatting and remove some micro optimizations that decrease readability

BenvdStouwe commented:
Hi Matty, 

I addressed your feedback with the latest commit. 

I also made a source generator that builds builders on compiletime. It's awesome. Unfortunately I can't open source that right now. I do hope that I'll be able to make it so in the near future.

Mattias1 reviewed (approved):
Looking good. Thank you for the PR, one step in the way to making this run on all platforms :)

Mattias1 commented:
> I also made a source generator that builds builders on compiletime. It's awesome. Unfortunately I can't open source that right now. I do hope that I'll be able to make it so in the near future.

I'd like to see (and ~~steal~~ use) that ;)
